pkgs/applications/networking/cluster/nomad/0.12.nix

```diff
@@ -0,0 +1,41 @@
+{ stdenv, buildGoPackage, fetchFromGitHub }:
```

Comment:
This file looks identical to the 0.11 derivation, except for the version/hash. In such cases, it's common to factor out the derivation to a file `generic.nix`, which takes the version and hash as additional arguments. So, `generic.nix` could start along the lines of:

```nix
{ stdenv
, buildGoPackage
, fetchFromGitHub

, version
, sha256
}:

buildGoPackage rec {
  inherit version;

  pname = "nomad";
  rev = "v${version}";

  goPackagePath = "github.com/hashicorp/nomad";
  subPackages = [ "." ];

  src = fetchFromGitHub {
    inherit rev sha256;
    owner = "hashicorp";
    repo = pname;
  };

# ...
```

And then `0.12.nix` could be:

```nix
{ callPackage }:

callPackage ./generic.nix {
  version = "0.12.0";
  sha256 = "01y54f7cpk31dgqdyrp331qam0mq9kjsrzhszysz7mk08lq3d2w8";
}
```

Untested, but just to give the general idea.

Reply:
I would freeze the 0.11 and have the `generic.nix` use `buildGoModule` instead of `buildGoPackage`.

Reply:
Updated thanks to @maxeaubrey (sorry it took so long, I've been super busy)

Reply:
@jonringer We're sticking with buildGoPackage for now though, bc Nomad still ships with a package-based-build

Reply:
ah, okay

Reply:
I still like @danieldk recommendation of creating a generic.nix and passing `version` and `sha256`. Seems much more extensible, and only needs to be modified in one location.

Reply:
@jonringer that's basically what this PR does? (Unless I'm missing something)

Reply:
kind of, but the `versionMap` kind of deviates from what I was thinking. I was thinking of something closer to boost:

https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/boost/generic.nix
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/boost/1.70.nix

Although boost is very complicated, and this is probably more scaffolding than what's needed for nomad.

Ultimately, the generic expression shouldn't be aware of any versions. Probably the best way for me to explain it.

@danieldk 's example code above is pretty much what I was looking for

Reply:
We'll update it I guess - but as the package maintainer that honestly seems unnecessarily more complex given the size of the closure and the desire to support up to ~3 versions.

I mostly wish we had an actual standard for this, because both styles exist within nix, especially for smaller packages

Reply:
That's fair, but nix is pretty flexible as a language, and circumstances like these aren't super common.

Reply:
Updated to follow that instead.